[계산기] 문상휘 1차 수정본 제출합니다.

moonwhistle opened the pull request:
코드 리뷰에 있는 모든 사항 수정하였습니다 ! 

* 코드리뷰 Controller - inputView, outputView 관련된 부분에 대해서 답변을 남겼지만, 확실하지 않은 것 같아서 제 답변 한 번 보시고 피드백 해주시면 감사하겠습니다!

moonwhistle committed:
Feat: 사칙연산 메서드 생성

moonwhistle committed:
Feat: 사직연산 테스트 코드 작성

Calculator : 0으로 나누었을 때 오류 발생시키는 기능 추가
CalculatorTest: 올바른 값이 나올 때와, 잘못 된 값이 나올 때 테스트 메서드 추가

moonwhistle committed:
Feat: 상수 컨벤션 수정

moonwhistle committed:
Update README.md

moonwhistle committed:
Update README.md

moonwhistle committed:
Feat: README 수정

moonwhistle committed:
Merge remote-tracking branch 'origin/sanghui-step1' into sanghui-step1

# Conflicts:
#	README.md

moonwhistle committed:
Feat: 초기화

moonwhistle committed:
Feat: 문자열 계산기 구현

moonwhistle committed:
Feat: 문자열 계산기 Junit5 단위테스트 실행

moonwhistle committed:
Feat: 문자열 계산기 Junit5 단위테스트 개행, 컨벤션 수정

moonwhistle committed:
Feat: 문자열 계산기 Junit5 단위테스트 개행, 컨벤션 수정

sosow0212 commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,14 @@
+        CalculatorController calculatorController = new CalculatorController(inputView, outputView);
+        calculatorController.startCalculator();
+    }
+
```
불필요한 공백 있습니다~

moonwhistle replied:
공백 리펙터링 하였습니다 !

sosow0212 commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,14 @@
+    public static void main(String[] args) {
+        InputView inputView = new InputView();
+        OutputView outputView = new OutputView();
+        CalculatorController calculatorController = new CalculatorController(inputView, outputView);
```
1. Controller로 넘겨주는 방식과 Controller에서 선언하는 방식은 뭐가 다를까요?

moonwhistle replied:
우선, 저번에 시범 보여주셨던 것을 바탕으로 한 번 생각해보고 구현하였습니다.

Controller로 넘겨주는 경우, 객체를 직접 생성자에 주입해주기 때문에 Controller에서 구현한 InputView 혹은 OutputView를 보다 안정성 있게 사용할 수 있을 것 같다는 생각이 들었습니다. 또한, 메서드 수정 시 번거롭지 않게 InputView 혹은 OutputView 를 하나 하나 바꿀 필요가 없다는 생각이 들었습니다. 

제가 틀렸거나 다른 이유가 있다면 피드백 주시면 감사하겠습니다!!

sosow0212 commented on src/main/java/controller/CalculatorController.java:
```diff
@@ -0,0 +1,27 @@
+    public void startCalculator(){
+        String inputStrings = inputView.getString();
+        StringCalculator stringCalculator = new StringCalculator(inputStrings);
+        List<String> paresStrings = stringCalculator.parseStrings();
+        List<Integer> numberList = stringCalculator.changeStringsToNumbers(paresStrings);
```
네이밍에 자료형 말고 조금 더 적합한 네이밍을 사용해보면 어떨까요?

moonwhistle replied:
네이밍 컨벤션 리펙터링 하였습니다!

sosow0212 commented on src/main/java/domain/StringCalculator.java:
```diff
@@ -0,0 +1,65 @@
+
+public class StringCalculator {
+
+    private final String STANDARD_SPLIT_STRINGS = "[,|:]";
+    private final int INITIAL_NUMBER = 0;
+    private final int FIRST_NUMBER = 1;
+    private final int SECOND_NUMBER = 2;
+    private final int THIRD_NUMBER = 3;
+    private final int CUSTOM_STRING_SPLIT_NUMBER = 5;
+    private final String EXCEPTION_MINUS = "음수 입력 불가.";
+    private final String EXCEPTION_INCORRECT = "제대로 입력하세요.";
+    private final String CHECK_CUSTOM_OR_STANDARD = "/";
```
final 과 static final은 무엇이 다를까요!

moonwhistle replied:
final - 불변값(초기화 되면)
static final - 전역적 불변값(초기화 필요)

상수의 경우 static final 을 선언해야 한다는 것을 알았습니다. 리펙터링 하였습니다

sosow0212 commented on src/main/java/domain/StringCalculator.java:
```diff
@@ -0,0 +1,65 @@
+
+    private void validateMinusNumber(List<Integer> numberList) {
+        for (Integer realNumber : numberList) {
+            if (realNumber < 0) {
```
0도 상수인데 빼먹은 것 같네요 ~

moonwhistle replied:
상수 컨벤션 수정하였습니다!

sosow0212 commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Test
+    @DisplayName("잘못된 더하기 테스트")
+    void 잘못된_더하기_테스트() {
+        //given
+        final Calculator calculator = new Calculator();
+        final int num1 = 1;
+        final int num2 = 2;
+        final int unexpected = 4;
+
+        //when
+        final int actual = calculator.add(num1, num2);
+
+        //then
+        assertNotEquals(unexpected, actual);
+    }
```
잘못된 더하기는 예외 케이스를 다루는 것이 아니면 따로 테스트하지 않아도 될 것 같아요
이미 위에서 더하기 테스트가 검증됐다면, 해당 메서드는 정상작동 하는거니깐요

moonwhistle replied:
알겠습니다 ! 참고하여 다른 메서드들의 잘못된 테스트 삭제하겠습니다.

sosow0212 commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,153 @@
+class CalculatorTest {
+
+    @Test
+    @DisplayName("더하기 테스트")
+    void 더하기_테스트() {
```
둘 중 하나만 써도 될 것 같아요~

moonwhistle replied:
알겠습니다. @DisplayName 어노테이션 삭제하여 테스트 수정하였습니다!

sosow0212 reviewed (changes requested):
수고하셨습니다!

구조도 깔끔하고 테스트도 보기 좋습니다.
질문하신 것에 대해서 정답은 없고 팀 컨벤션에 따라 다를 수 있지만, 저는 개인적으로 참조하는 것이 최소한이 되도록 만드는게 구조적으로는 좋은 것 같습니다!

하지만, 저는 예를 들어 생성자에서 예외를 하고 추후 다른 메서드에서 값을 이용하는 경우라면 생성자 검증 테스트를 따로 하고, 다른 메서드에서 의존이 조금 있더라고 핵심기능일테니 그냥 하는 편이긴 합니다.
하지만, 최대한 의존하지 않도록 노력하는 편입니다!

두 번째 질문에 대해서는 assertEqauls 말고 어떤 검증 메서드가 있는지 확인해보면 좋을 것 같습니다.
지금도 나쁘지는 않지만, 체이닝으로 한 번에 할 수도 있지 않을까요? 이건 느끼는게 다 다를 수 있어서 직접 찾아보고 비교해보면 좋을 것 같습니다!

마지막 개행이 계속 추가되는 것 같은데 이 부분만 한 번 확인해주세요!
수고하셨습니다 👍

moonwhistle committed:
Refactor: 컨벤션 수정 - 마지막 개행 삭제

moonwhistle committed:
Refactor: 컨벤션 수정 - 네이밍 수정

moonwhistle committed:
Refactor: 컨벤션 수정 - 상수 컨벤션 수정

moonwhistle committed:
Refactor: 컨벤션 수정 - 상수 컨벤션 수정

moonwhistle committed:
Refactor: 테스트 수정

올바른 경우의 테스트가 검증된 메서드의 잘못된 경우의 테스트 삭제, @Display 어노테이션으로 인해 테스트 메서드 이름과 겹치는 경우 @Display 어노테이션 삭제

moonwhistle commented:
코드 리뷰에 있는 모든 사항 수정하였습니다 ! 

코드리뷰 Controller - inputView, outputView 관련된 부분에 대해서 답변을 남겼지만, 확실하지 않은 것 같아서 제 답변 한 번 보시고 피드백 해주시면 감사하겠습니다!

sosow0212 reviewed (approved):
구조도 잘 나눠주셨고, 처음인데 잘 나눠주셨네요!!
해당 미션 리뷰에서는 구조와 컨벤션 위주로 남겨봤어요~ 다음 미션부터는 구조적으로 같이 더 깊게 공부하면 좋을 것 같습니다.

'객체지향의 사실과 오해' 책도 꼭 읽어보시면 좋겠습니다!
추가적으로 궁금한 점 있다면 카톡으로 남겨주세요~
